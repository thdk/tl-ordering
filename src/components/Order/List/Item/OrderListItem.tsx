import * as React from "react";

import { connect } from "react-redux";
import { IState } from "../../../../core/app/types";
import { IOrder } from "../../../../core/orders/types";
import OrderLink from "../../../links/OrderLink";

interface IOrderListItemProps {
    order: IOrder;
}

interface IPropsFromState { total: number; }

type Props = IOrderListItemProps & IPropsFromState;

const OrderListItem = (props: Props) => {
    const { order: { id, customerId }, total } = props;

    return (
        <div className="row">
            <div className="row-left">
                <div>{id}</div>
                <div>{customerId}</div>
            </div>
            <div className="row-right">
                <div>{total.toFixed(2)}</div>
                <div>
                    <OrderLink orderId={id}>detail</OrderLink>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state: IState, ownProps: IOrderListItemProps) => {

    const total = state.orderItems.byOrderId[ownProps.order.id].reduce((p, c) => {
        const product = state.products.byId[c.productId];
        if (!product) { console.error("No product found for " + c.productId + ". Using '0' as price"); }
        p += (product ? product.price : 0) * c.quantity;
        return p;
    }, 0);

    return { total };
};

const ConnectedOrderListItem = connect(
    mapStateToProps,
)(OrderListItem);

export default ConnectedOrderListItem;
