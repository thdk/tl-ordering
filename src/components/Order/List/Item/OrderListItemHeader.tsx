import * as React from "react";

export const OrderListItemHeader: React.FunctionComponent = () => {
    return (
        <div className="row header">
            <div className="row-left">
                <div>Order id</div>
                <div>Customer id</div>
            </div>
            <div className="row-right">
                <div>Total</div>
                <div>
                    Order detail
                </div>
            </div>
        </div>
    );
};

OrderListItemHeader.displayName = "OrderListItemHeader";

export default OrderListItemHeader;
