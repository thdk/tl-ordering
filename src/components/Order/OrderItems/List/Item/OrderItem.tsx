import React from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import { removeOrderItem } from "../../../../../actions/orderItems";
import { IOrderItem } from "../../../../../interfaces/orders";
import { IState } from "../../../../../interfaces/state";

interface IOrderItemPropsFromState { unitPrice: number; }
interface IDispatchFromProps { onDeleteOrderItem: (productId: string) => void; }
export interface IOrderItemProps extends IOrderItem {
    orderId: string;
}

type Props = IOrderItemPropsFromState & IDispatchFromProps & IOrderItemProps;

const OrderItem = (props: Props) => {
    const { onDeleteOrderItem, productId, quantity, unitPrice } = props;

    const total = (quantity * unitPrice).toFixed(2);
    return (
        <li>
            {productId}: {quantity} x ${unitPrice} = ${total}
            <input type="button" value="x" onClick={() => onDeleteOrderItem(productId)} />
        </li>
    );
};

const mapDispatchToProps = (dispatch: Dispatch, ownProps: IOrderItemProps) => {
    return {
        onDeleteOrderItem: (productId: string) => {
            dispatch(removeOrderItem(productId, ownProps.orderId));
        },
    };
};

const mapStateToProps = (state: IState, ownProps: IOrderItemProps) => {
    return {
        unitPrice: state.products.byId[ownProps.productId].unitPrice,
    };
};

const ConnectedOrderItem = connect
    <IOrderItemPropsFromState, IDispatchFromProps, IOrderItemProps, IState>(
        mapStateToProps,
        mapDispatchToProps,
    )(OrderItem);

export default ConnectedOrderItem;
