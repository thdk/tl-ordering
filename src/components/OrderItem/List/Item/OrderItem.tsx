import React, { useCallback } from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";

import { IState } from "../../../../core/app/types";
import { removeOrderItem } from "../../../../core/orderitems/actions";
import { IOrderItem } from "../../../../core/orderitems/types";
import { getProduct } from "../../../../core/products/reducer";

export interface IOrderItemProps extends IOrderItem {
    orderId: string;
}

type Props = IOrderItemProps & DispatchProps & StateProps;

const OrderItem: React.FunctionComponent<Props> = (props: Props) => {
    const { onDeleteOrderItem, productId, quantity, product } = props;

    const onClick = useCallback(() => onDeleteOrderItem(productId), [productId]);

    if (!product) { return null; }

    const { description, price } = product;
    return (
        <li>
            <span className="orderitem-product-description">
                {description}
            </span>:
            <span className="orderitem-quantity">
                {quantity}
            </span> x ${price.toFixed(2)} = ${(price * quantity).toFixed(2)}
            <input type="button" value="x" onClick={onClick} />
        </li>
    );
};

OrderItem.displayName = "OrderItem";

type StateProps = ReturnType<typeof mapStateToProps>;

const mapStateToProps = (state: IState, {productId}: IOrderItemProps) => ({
    product: getProduct(state, productId),
});

type DispatchProps = ReturnType<typeof mapDispatchToProps>;

const mapDispatchToProps = (dispatch: Dispatch, ownProps: IOrderItemProps) => ({
    onDeleteOrderItem: (productId: string) => {
        dispatch(removeOrderItem(productId, ownProps.orderId));
    },
});

const ConnectedOrderItem = connect(
    mapStateToProps,
    mapDispatchToProps,
)(OrderItem);

export default ConnectedOrderItem;
