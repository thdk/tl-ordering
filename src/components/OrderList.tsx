import React, { useEffect } from 'react'
import { ThunkDispatch } from 'redux-thunk';
import { connect } from 'react-redux'

import { IOrder } from '../interfaces/orders';
import OrderLink from './links/OrderLink';
import Order from './Order';
import { IState } from '../interfaces/state';
import { placeOrder, fetchOrders } from '../actions/orders';

type PropsFromState = {
    orders: IOrder[];
    isLoading: boolean;
};

type PropsFromDispatch = {
    fetchOrders: () => void;
}

type Props = PropsFromState & PropsFromDispatch;

const OrderList = (props: Props) => {
    const { fetchOrders, orders, isLoading } = props;

    useEffect(() => {
        // fetchOrders()
    }, []);

    return isLoading
        ? <div>LOADING</div>
        : <ul>
            {orders.map(order => {
                const { id } = order;
                return <React.Fragment key={id}>
                    <OrderLink orderId={id}>Order detail</OrderLink>
                    <Order key={id} id={id} />
                </React.Fragment>;
            })}
        </ul>;
};

const mapStateToProps = (state: IState): PropsFromState => {
    return {
        orders: state.orders.orders,
        isLoading: state.orders.isLoading
    }
};

const mapDispatchToProps = (dispatch: ThunkDispatch<{}, {}, any>): PropsFromDispatch => {
    return {
        fetchOrders: () => dispatch(fetchOrders()),
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(OrderList);