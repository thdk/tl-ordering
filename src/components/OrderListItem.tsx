import * as React from 'react';

import { IOrder } from "../interfaces/orders";
import OrderLink from './links/OrderLink';
import { IState } from '../interfaces/state';
import { connect } from 'react-redux';

export interface IOrderListItemProps {
    order: IOrder;
}

export type PropsFromState = { total: number };

type Props = IOrderListItemProps & PropsFromState;

const OrderListItem = (props: Props) => {
    const { order: { id, customerId }, total } = props;

    return (
        <div className="row">
            <div className="row-left">
                <div>{id}</div>
                <div>{customerId}</div>
            </div>
            <div className="row-right">
                <div>{total.toFixed(2)}</div>
                <div>
                    <OrderLink orderId={id}>detail</OrderLink>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state: IState, ownProps: IOrderListItemProps) => {

    const total = state.orderItems.byOrderId[ownProps.order.id].reduce((p, c) => {
        const product = state.products.byId[c.productId];
        if (!product) console.error("No product found for " + c.productId + ". Using '0' as price");
        p += (product ? product.unitPrice : 0) * c.quantity;
        return p;
    }, 0);

    return { total };
};

const ConnectedOrderListItem = connect(
    mapStateToProps
)(OrderListItem);

export default ConnectedOrderListItem;