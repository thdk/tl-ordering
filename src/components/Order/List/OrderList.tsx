import React from "react";
import { connect } from "react-redux";

import { IState } from "../../../core/app/types";
import { getVisibleOrders } from "../../../core/orders/reducer";
import { OrderListItem } from "./Item/OrderListItem";
import OrderListItemHeader from "./Item/OrderListItemHeader";

type OrderListProps = StateProps;

export const OrderList = (props: OrderListProps) => {
    const { orders } = props;

    const listJSX = orders.length
        ? <div className="list">
            <OrderListItemHeader></OrderListItemHeader>
            {orders.map(order => <OrderListItem key={order.id} order={order}></OrderListItem>)}
        </div>
        : <div>
            No orders found.
            <br />
            <a href="?debug">Use mocked api?</a>
        </div>;

    return listJSX;
};

type StateProps = ReturnType<typeof mapStateToProps>;

const mapStateToProps = (state: IState) => ({
    orders: getVisibleOrders(state),
});

export default connect(mapStateToProps)(OrderList);
