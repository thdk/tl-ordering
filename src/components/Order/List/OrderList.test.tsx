import { shallow, ShallowWrapper } from "enzyme";
import React from "react";

import { OrderListItem } from "./Item/OrderListItem";
import { OrderListItemHeader } from "./Item/OrderListItemHeader";
import { mapStateToProps, OrderList } from "./OrderList";

describe("Render OrderList", () => {
    let component: ShallowWrapper;
    describe("with multipe orders", () => {

        beforeEach(() => {
            component = shallow(
                <OrderList
                    orders={[
                        {
                            customerId: "c1",
                            id: "o1",
                            total: 5,
                            items: [],
                        },
                        {
                            customerId: "c2",
                            id: "o2",
                            total: 5,
                            items: [],
                        },
                        {
                            customerId: "c2",
                            id: "o3",
                            total: 5,
                            items: [],
                        },
                    ]} />,
            );
        });

        it("should match snapshot", () => {
            expect(component.debug()).toMatchSnapshot();
        });

        it("should have OrderListItemHeader component", () => {
            const listHeader = component.find(OrderListItemHeader);
            expect(listHeader.exists()).toBe(true);
        });

        it("should have a OrderListItem component for each order", () => {
            const items = component.find(OrderListItem);
            expect(items.length).toBe(3);
        });

        it("should display orders in correct order", () => {
            const items = component.find(OrderListItem);
            const orderIds = items.map(o => o.props().order.id);

            expect(orderIds).toEqual(["o1", "o2", "o3"]);
        });
    });

    describe("without orders", () => {

        beforeEach(() => {
            component = shallow(
                <OrderList
                    orders={[]}
                />,
            );
        });

        it("should match snapshot", () => {
            expect(component.debug()).toMatchSnapshot();
        });
    });
});

describe("OrderList mapStateToProps", () => {
    const stateProps = mapStateToProps({
        orderItems: {
            byOrderId: {},
        },
        orders: {
            byId: {
                o1: {
                    customerId: "c1",
                    id: "o1",
                },
            },
            isLoading: false,
            visibleIds: ["o1"],
        },
        products: {
            allIds: [],
            byId: {},
            isLoading: false,
        },
    });

    expect(stateProps.orders.length).toBe(1);
});
