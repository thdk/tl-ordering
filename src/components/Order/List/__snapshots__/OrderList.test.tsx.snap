// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render OrderList with multipe orders should match snapshot 1`] = `
"<div className=\\"list\\">
  <OrderListItemHeader />
  <OrderListItem order={{...}} />
  <OrderListItem order={{...}} />
  <OrderListItem order={{...}} />
</div>"
`;

exports[`Render OrderList without orders should match snapshot 1`] = `
"<div>
  No orders found.
  <br />
  <a href=\\"?debug\\">
    Use mocked api?
  </a>
</div>"
`;
